docs/enhancements/dynamic-fetching-of-openshift-versions.md

```diff
@@ -0,0 +1,290 @@
+  a total of ~140 versions).
+
+### Non-Goals
+
+* Dynamic handling of y-versions, live-iso discovery images, and such. A new minor
+  OpenShift version is introduced every 4 months, requires additional code changes
+  (for example in feature-support-level list) and it isn't worth the costs and
+  implications of making assisted-image-service refresh itself and pull-in new
+  images.
+* Handling OpenShift versions in a kubernetes API interface. For this use-case
+  versions are externally configured in the ``ClusterImageSet`` custom resource.
+
+## Proposal
+
+The assisted-service will make use of a new configuration variable called
+``RELEASE_SOURCES``, which outlines the supported matrix. For example:
+
+```
+RELEASE_SOURCES='{
+   "4.11": {
+      "upgrade_channels": [
+         {
+            "cpu_architecture": "x86_64",
+            "channels": ["stable"]
+         }
+      ]
+   },
+   "4.12": {
+      "upgrade_channels": [
+         {
+            "cpu_architecture": "x86_64",
+            "channels": ["candidate", "stable"]
+         },
+         {
+            "cpu_architecture": "arm64",
+            "channels": ["candidate", "stable"]
+         },
+         {
+            "cpu_architecture": "multi",
+            "channels": ["candidate", "stable"]
+         }
+      ]
+   }
+}'
+```
+
+The assisted-service periodically polls versions from the upgrade-channels API
+in the form of ``https://api.openshift.com/api/upgrades_info/v1/graph?channel=<channel>&arch=<arch>``.
+For example:
+```
+$ curl -s 'https://api.openshift.com/api/upgrades_info/v1/graph?channel=stable-4.11&arch=amd64' \
+    | jq -r '.nodes[] | select(.version | startswith("4.11")) | .version + " => " + .payload'
+4.11.9 => quay.io/openshift-release-dev/ocp-release@sha256:94b611f00f51c9acc44ca3f4634e46bd79d7d28b46047c7e3389d250698f0c99
+4.11.4 => quay.io/openshift-release-dev/ocp-release@sha256:e04ee7eea9e6a6764df218bcce3ee64e1816445158a0763954b599e0fc96580b
+...
+```
+
+When an API consumer (either it's the UI or a direct HTTP consumer) requests
+for the available versions via ``/v2/openshift-versions``, versions will be read
+from the upgrades-channels API if it weren't queried for a configurable amount
+of time (for example, an hour).
+
+To help fetching only certain versions via auto-completion like U/X, we should
+provide a parameter ``version_pattern`` for the ``/v2/openshift-versions``
+endpoint, along with a ``only_latest`` parameter. For example:
+```
+/v2/openshift-versions:
+   get:
+      ...
+      parameters:
+        - in: query
+          name: version_pattern
+          description: Version pattern for the returned releases. E.g. '4.12'.
+          type: string
+          required: false
+        - in: query
+          name: only_latest
+          description: It true, returns only the latest version for each minor.
+          type: boolean
+          allowEmptyValue: true
+          default: false
```

Comment:
@avishayt @danielerez do you think it should be ``true`` by default?
otherwise it returns a big list that might be a bit weird for some API consumers

Reply:
Yeah, changing the current behaviour of the API might be confusing for existing users.
What about introducing instead a property for getting all versions? Something like 'include_all_minor' boolean.
From the UI perspective, it seems that only two variants will be used: the old behaviour (only latest) and version_pattern. So getting all minor versions seems like a rare option that deserves a separate special param.

Reply:
``include_all_minor`` is the same as having the proposed ``only_latest``, no?

Reply:
> `include_all_minor` is the same as having the proposed `only_latest`, no?

Yeah, just the opposite :) Not crucial, just sounds more straightforward than 'only_latest=false'

Reply:
hmm, by the way [minor is the y part](https://semver.org/#summary) so do you mean ``include_all_patch_versions``?

Reply:
> hmm, by the way [minor is the y part](https://semver.org/#summary) so do you mean `include_all_patch_versions`?

right, tried to shorten 'include all for minor', naming is hard:) maybe can be just 'all_versions' and explain in the description what is returned exactly.

Reply:
Sounds good
Thanks!